FWCore/Framework/interface/ESProducer.h

```diff
@@ -90,6 +90,46 @@ Example: two algorithms each creating only one objects
 // forward declarations
 namespace edm {
   namespace eventsetup {
+    namespace detail {
+      template <typename F>
+      struct RetArgTypeImpl;
+
+      // function pointer
+      template <typename R, typename T>
+      struct RetArgTypeImpl<R (*)(T const&)> {
+        using argument_type = T;
+        using return_type = R;
+      };
+
+      // mutable functor/lambda
+      template <typename R, typename T, typename O>
+      struct RetArgTypeImpl<R (O::*)(T const&)> {
+        using argument_type = T;
+        using return_type = R;
+      };
+
+      // const functor/lambda
+      template <typename R, typename T, typename O>
+      struct RetArgTypeImpl<R (O::*)(T const&) const> {
+        using argument_type = T;
+        using return_type = R;
+      };
+
+      template <typename F, typename = void>
+      struct RetArgType;
+
+      template <typename F>
+      struct RetArgType<F, std::enable_if_t<std::is_class_v<F>>> {
+        using argument_type = typename RetArgTypeImpl<decltype(&F::operator())>::argument_type;
+        using return_type = typename RetArgTypeImpl<decltype(&F::operator())>::return_type;
+      };
+
+      template <typename F>
+      struct RetArgType<F, std::enable_if_t<std::is_pointer_v<F>>> {
+        using argument_type = typename RetArgTypeImpl<F>::argument_type;
+        using return_type = typename RetArgTypeImpl<F>::return_type;
+      };
+    }  // namespace detail
```

Comment:
There might be a better place for these metafunctions somewhere else.

Reply:
Yeah, this probably should be moved to its own file.

Reply:
Moved to `FWCore/Framework/interface/es_impl` and renamed the metafunctions to `ArgumentReturnTypes`